src/plan.rs

```diff
@@ -0,0 +1,1251 @@
+                Some(can_sign.taproot.sig_len())
+            }
+        })
+    }
+
+    pub(crate) fn has_taproot_script_key_hash(
+        &self,
+        key: &hash160::Hash,
+        tap_leaf_hash: &TapLeafHash,
+    ) -> Option<usize> {
+        self.keys.get(key).and_then(|(_, can_sign)| {
+            if !can_sign.taproot.script_spend.is_available(tap_leaf_hash) {
+                None
+            } else {
+                Some(can_sign.taproot.sig_len())
+            }
+        })
+    }
+}
+
+impl AssetProvider<DefiniteDescriptorKey> for Assets {
+    fn lookup_ecdsa_sig(&self, pk: &DefiniteDescriptorKey) -> bool {
+        self.has_ecdsa_key(pk)
+    }
+
+    fn lookup_tap_key_spend_sig(&self, pk: &DefiniteDescriptorKey) -> Option<usize> {
+        self.has_taproot_internal_key(pk)
+    }
+
+    fn lookup_tap_leaf_script_sig(
+        &self,
+        pk: &DefiniteDescriptorKey,
+        tap_leaf_hash: &TapLeafHash,
+    ) -> Option<usize> {
+        self.has_taproot_script_key(pk, tap_leaf_hash)
+    }
+
+    fn lookup_raw_pkh_pk<Ctx: ScriptContext>(&self, hash: &hash160::Hash) -> Option<usize> {
+        self.keys.get(hash).map(|(pk, _)| Ctx::pk_len(pk))
+    }
+
+    fn lookup_raw_pkh_ecdsa_sig<Ctx: ScriptContext>(&self, hash: &hash160::Hash) -> Option<usize> {
+        self.keys.get(hash).map(|(pk, _)| Ctx::pk_len(pk))
+    }
+
+    fn lookup_raw_pkh_tap_leaf_script_sig(
+        &self,
+        (key, lh): &(hash160::Hash, TapLeafHash),
+    ) -> Option<usize> {
+        self.has_taproot_script_key_hash(key, lh)
+    }
+
+    fn lookup_sha256(&self, hash: &sha256::Hash) -> bool {
+        self.sha256_preimages.contains(hash)
+    }
+
+    fn lookup_hash256(&self, hash: &hash256::Hash) -> bool {
+        self.hash256_preimages.contains(hash)
+    }
+
+    fn lookup_ripemd160(&self, hash: &ripemd160::Hash) -> bool {
+        self.ripemd160_preimages.contains(hash)
+    }
+
+    fn lookup_hash160(&self, hash: &hash160::Hash) -> bool {
+        self.hash160_preimages.contains(hash)
+    }
+
+    fn check_older(&self, s: Sequence) -> bool {
+        if let Some(rt) = &self.relative_timelock {
+            return rt.is_relative_lock_time()
+                && rt.is_height_locked() == s.is_height_locked()
+                && s <= *rt;
+        }
+
+        false
+    }
+
+    fn check_after(&self, l: LockTime) -> bool {
+        if let Some(at) = &self.absolute_timelock {
+            let cmp = l.partial_cmp(at);
+            return cmp == Some(Ordering::Less) || cmp == Some(Ordering::Equal);
+        }
+
+        false
+    }
+}
+
+impl FromIterator<DescriptorPublicKey> for Assets {
+    fn from_iter<I: IntoIterator<Item = DescriptorPublicKey>>(iter: I) -> Self {
+        Assets {
+            keys: iter
+                .into_iter()
+                .map(|pk| {
+                    (
+                        pk.clone()
+                            .at_derivation_index(0)
+                            .to_pubkeyhash(SigType::Ecdsa),
```

Comment:
Ahh I think I am confused by public key hashes and miniscript again.

Why is this indexing by the pubkeyhash of the thing at derivation_index 0? I mean if the thing you need to sign is the pubkey hash at derivation index 7 how are you going to be able to look it up via this index.

Can we make pubkeyhashes a separate concern in a different field where you say what pubkeyhashes you can sign for and nothing else. i.e. just a `HashSet<hash160::Hash>`.

Reply:
If we ignore pubkey hashes entirely then the satisfaction code will not support satisfying parsed scripts anymore, will that be ok?

Reply:
You mean, it won't support satisfying parsed scripts that have pubkey hashes in them? I think that's OK, at least for the first iteration of this API.

If you mean that we can't satisfy parsed scripts *at all* we should fix this now.

Reply:
> If we ignore pubkey hashes entirely then the satisfaction code will not support satisfying parsed scripts anymore, will that be ok?

Parsed scripts with raw hashes are not miniscripts or miniscript descriptors. We should not be creating plans for incomplete/partial descriptors. If we have access to full descriptors, like `pkh(KEY)`, this will work as expected. 

In the future, this will be supported cleanly when we have support for partial descriptors.